compiler/damlc/daml-visual/src/DA/Daml/Visual.hs

```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
 addCreateChoice TemplateChoices {..} lookupData = nodeIdForChoice lookupData tplNameCreateChoice
     where tplNameCreateChoice = LF.ChoiceName $ T.pack $ DAP.renderPretty (headNote "addCreateChoice" (LF.unTypeConName (LF.tplTypeCon template))) ++ "_Create"
 
-constructSubgraphsWithLables :: Map.Map LF.ChoiceName ChoiceDetails -> TemplateChoices -> SubGraph
-constructSubgraphsWithLables lookupData tpla@TemplateChoices {..} = SubGraph nodesWithCreate template
+-- This is copied from PrettyScenarios but depending on SS for visual seems odd
+labledField :: T.Text -> T.Text -> T.Text
+labledField fname "" = fname
+labledField fname label = fname <> "." <> label
+
+typeConFieldsNames :: LF.World -> (LF.FieldName, LF.Type) -> [T.Text]
+typeConFieldsNames world (LF.FieldName fName, LF.TConApp tcn _) = map (labledField fName) (typeConFields tcn world)
+typeConFieldsNames _ (LF.FieldName fName, _) = [fName]
+
+typeConFields :: LF.Qualified LF.TypeConName -> LF.World -> [T.Text]
+typeConFields qName world = case LF.lookupDataType qName world of
```

Comment:
It looks like currently you simply flatten all nested records into a single list of fields. Can we display them in a more hierarchical way? Throwing away the structure that the user created doesn’t seem like a nice way to present this.
I could also see this blowing up on large records if you recurse without a limit. Maybe we should make this configurable, e.g., initially have a single parameter for recursing on fields where `0` does not display any fields in the graph (i.e. the behavior without this PR), 1 goes down a single level, 2 recurses once, …

Reply:
Oh I see, I missed the dots. Thanks for the explanation. I’m still a bit worried that this doesn’t scale to large nested records so I think the limit is probably still a good idea but I’m fine with that being added separately.